<a href="https://colab.research.google.com/github/stephyi/10Academy/blob/master/002_SQL_Basics_for_Data_Science_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
%load_ext sql
%sql sqlite://

'Connected: @None'

In [ ]:
accidents = pd.read_csv('traffic_accidents.csv')
%sql DROP TABLE if EXISTS accidents;
%sql PERSIST accidents;

 * sqlite://
Done.
 * sqlite://


'Persisted accidents'

In [ ]:
# Defining data

%sql SELECT * FROM accidents LIMIT 5

 * sqlite://
Done.


[(0, 758369, 'JC469557', '2019-10-12T01:57:00', 1, 'DRIVER', None, 720366, None, 'DODGE', 'NITRO', 'IL', 2011, 9, 'UNKNOWN', 'PASSENGER', 'UNKNOWN/NA', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (1, 758356, 'JC469523', '2019-10-12T00:30:00', 1, 'DRIVER', None, 720356, None, 'VOLVO', 'XC90', 'IL', 2020, 0, 'NONE', 'SPORT UTILITY VEHICLE (SUV)', 'PERSONAL', 'S', 'SLOW/STOP IN TRAFFIC', None, None, 1, None, None, None, None, None, None, 1, None, None, None, None, None, None, None, None, None, None, 'SIDE-RIGHT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (2, 758329, 'JC469467', '2019-10-11T23:41:00', 1, 'DRIVER', 2.0, 720325, None, 'TOYOTA MOTOR COMPANY, LTD.', 'UNKNOWN', 'IL', 2015, 5, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'PASSING/OVERTAKING', 1, None, 3, None, 'unknown', 'unknown', None, None, None, 1, 1, 1, 1, 1, None, None, None, None, None, None, 'REAR', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (3, 758330, 'JC469467', '2019-10-11T23:41:00', 2, 'DRIVER', None, 720334, None, 'SUBARU', 'FORESTER', 'IL', 2014, 6, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'TURNING LEFT', None, None, 1, None, None, None, None, 1, 1, None, None, None, None, None, None, None, None, None, 1, None, 'FRONT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (4, 758333, 'JC469482', '2019-10-11T23:20:00', 1, 'DRIVER', None, 720328, None, 'CHEVROLET', 'IMPALA', 'IL', 2017, 3, 'UNKNOWN', 'PASSENGER', 'PERSONAL', 'N', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'UNKNOWN', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None)]

In [ ]:
# 1. Display a list of 5 makes that had the highest number of crashes?

%sql SELECT make, model, count(make) FROM accidents GROUP BY make ORDER BY count(make) desc LIMIT 5;

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 'COROLLA', 637),
 ('CHEVROLET', 'CRUZE', 572),
 ('FORD', 'FUSION', 565),
 ('NISSAN', 'NISSAN ALTIMA', 491),
 ('HONDA', 'Odyssey (minivan)', 369)]

The list of the 5 makes that had the highest number of crashes is topped by Toyota Motor, most likely because it has the most cars on the roads followed by chevrolet, Ford, Nissan and Honda respectively.

In [ ]:
#@title Default title text
# 2. Most personal vehicles involved in crashes were travelling towards which direction?

%sql SELECT MAX("travel_direction"), count("travel_direction") FROM accidents WHERE ("vehicle_use") = "PERSONAL"

 * sqlite://
Done.


[('W', 3562)]

Most of the personal vehicles that were involved in crashes were travelling to the West, most likely on the same highway.

In [ ]:
# 3. How many passengers were involved in personal cars were in accidents?

%sql SELECT "vehicle_use", count("occupant_cnt") FROM accidents WHERE ("vehicle_use") = "PERSONAL"

 * sqlite://
Done.


[('PERSONAL', 3562)]

A total of 3562 passengers in personal cars and travelling to the west were involved in accidents.

In [ ]:
# 4. How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?

%sql SELECT "vehicle_type", count("occupant_cnt") FROM accidents WHERE ("vehicle_use") = "PERSONAL" AND ("vehicle_type") = 	"SPORT UTILITY VEHICLE (SUV)"

 * sqlite://
Done.


[('SPORT UTILITY VEHICLE (SUV)', 529)]

A total of 529 passengers travelling in personal Sport Utility vehicles were involved in accidents.

In [ ]:
# 5. Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.

%sql SELECT count("MERCEDES-BENZ"), maneuver FROM accidents WHERE maneuver = "PARKED"

 * sqlite://
Done.


[(708, 'PARKED')]

708 mercedes-benz cars were involved in accidents while they were parked.

In [ ]:
# 6. Compute the total occupant count per vehicle count.

%sql SELECT SUM("occupant_cnt"), count("crash_unit_id") FROM accidents

 * sqlite://
Done.


[(5671, 5040)]

In all the accidents which occured, there were 5040 different accidents which occured with 5671 passagers involved in these accidents.

In [ ]:
# 7. Sort the number of passengers in descending order by travel direction taking into account travel direction. Return travel direction and no. of passengers.

%sql SELECT "travel_direction", SUM("occupant_cnt") FROM accidents  GROUP BY "travel_direction" ORDER BY SUM("occupant_cnt") DESC;

 * sqlite://
Done.


[('N', 1405),
 ('W', 1325),
 ('S', 1316),
 ('E', 1219),
 ('UNKNOWN', 118),
 ('SE', 92),
 ('NW', 83),
 ('SW', 65),
 ('NE', 48)]

Most of the passengers (1405) who were involved in road traffic accidents were travelling to the northern direction followed by the west with 1325 passengers and the South with 1316 passengers.

In [ ]:
# 8. Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre. 

%sql SELECT model, maneuver, count("model") FROM accidents WHERE "vehicle_defect" = "NONE" GROUP BY "model" ORDER BY count("model") DESC LIMIT 5

 * sqlite://
Done.


[('UNKNOWN', 'STRAIGHT AHEAD', 403),
 ('CAMRY', 'TURNING RIGHT', 135),
 ('NISSAN ALTIMA', 'SLOW/STOP IN TRAFFIC', 82),
 ('SENTRA (DATSUN AND NISSAN HAVE MERGED)', 'STRAIGHT AHEAD', 72),
 ('MALIBU (CHEVELLE)', 'PARKED', 69)]

403 vehicles of unknown model were involved in road accidents while travelling straight ahead, they didn't have any defects and they drove away which is why their models are unknown, these are followed by 135 incidents of Camrys which were turning right, 82 Nissan Altimas which were slow/ coming to a stop indicating an existing defect with their braking systems, 72 sentra's travelling strainght ahead and lastly 69 malibu(chevelle) which were parked.

In [ ]:
# 9. Which ridesharing vehicles in an accident had the most number of passengers? Return model, and no of passengers.

%sql SELECT model, SUM("occupant_cnt") FROM accidents WHERE "vehicle_use" = "RIDESHARE SERVICE" GROUP BY model ORDER BY SUM("occupant_cnt") DESC

 * sqlite://
Done.


[('Prius', 11),
 ('UNKNOWN', 10),
 ('CAMRY', 9),
 ('COROLLA', 9),
 ('NISSAN ALTIMA', 9),
 ('MKZ', 8),
 ('ROGUE', 7),
 ('MALIBU (CHEVELLE)', 5),
 ('PASSAT', 5),
 ('FUSION', 4),
 ('JETTA', 4),
 ('SENTRA (DATSUN AND NISSAN HAVE MERGED)', 4),
 ('SONATA', 4),
 ('Aveo', 3),
 ('C300', 3),
 ('FORTE', 3),
 ('Focus', 3),
 ('Outlander', 3),
 ('RX350', 3),
 ('SOUL', 3),
 ('Accent', 2),
 ('CARAVAN', 2),
 ('CIVIC', 2),
 ('CRUZE', 2),
 ('ELANTRA', 2),
 ('ES300', 2),
 ('GENESIS', 2),
 ('Highlander(beginning vehicle year 2001)', 2),
 ('JOURNEY', 2),
 ('MDX', 2),
 ('MKT', 2),
 ('TUCSON', 2),
 ('323', 1),
 ('ACURA', 1),
 ('CRV', 1),
 ('FORD', 1),
 ('GRAND MARQUIS', 1),
 ('Grand Cherokee (for vehicle years 1989 to present)', 1),
 ('Mazda 3', 1),
 ('Murano', 1),
 ('NAVIGATOR', 1),
 ('S90', 1),
 ('SORENTO', 1),
 ('ACCORD', 0)]

The prius is the number one ride sharing vehicle that had the most number of pasengers which was involved in an accident, followed by several unknown models and the Camry came third.